### PR TITLE
feat: Add Create Deal for Data Market Module

### PR DESCRIPTION
**Pull request by inchori**
- Configure proto file and ```app/app.go``` for ```x/market``` module.
- Implement Create Deal feature for ```x/market``` module.
- Complete Test Code for ```keeper/deal_test.go```, ```types/deal_test.go```, and ```market/genesis_test.go```.
- Use Hash function of Cosmos SDK v0.43.0 for making Deal Address.

TODO
- Tx Cmd for Create Deal
- Validation for MsgCreateDeal
- Implement ExportGenesis function for ```x/market``` module.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,9 +1,10 @@
-FROM golang:1.16.5-buster AS build-env
```
Comment (gyuguen) [inline Dockerfile:1]: 이 부분은 이전 개발 소스로 보여지는대 뭔가 새로 추가된게 있나요?

File: app/app.go [hunk 2]
```diff
@@ -248,14 +252,15 @@ type App struct {
...
-	didKeeper   didkeeper.Keeper
-	burnKeeper  burnkeeper.Keeper
-	tokenKeeper tokenkeeper.Keeper
-	wasmKeeper  wasm.Keeper
+	aolKeeper    aolkeeper.Keeper
+	didKeeper    didkeeper.Keeper
+	burnKeeper   burnkeeper.Keeper
+	tokenKeeper  tokenkeeper.Keeper
+	wasmKeeper   wasm.Keeper
+	MarketKeeper marketkeeper.Keeper
```
Comment (gyuguen) [inline app/app.go:260]:
```suggestion
	marketKeeper marketkeeper.Keeper
```
다른 keeper와 마찬가지가지로 소문자로 가면 어떨까요

File: app/test_helper.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	return app
+}
+
+func FundAccount(bankKeeper bankkeeper.Keeper, ctx sdk.Context, addr sdk.AccAddress, amounts sdk.Coins) error {
+	if err := bankKeeper.MintCoins(ctx, minttypes.ModuleName, amounts); err != nil {
+		return err
+	}
+
+	return bankKeeper.SendCoinsFromModuleToAccount(ctx, minttypes.ModuleName, addr, amounts)
+}
```
Comment (gyuguen) [inline app/test_helper.go:43]: 현재 사용하지는 않는거 같은대 나중에 사용하려고 하나요?

File: x/market/keeper/deal_test.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+		WantDataCount:        10000,
+		TrustedDataValidator: []string{acc2.String()},
+		Owner:                acc1.String(),
+	}
+
+	owner, _ := sdk.AccAddressFromBech32(tempDeal.GetOwner())
+
+	dealId, _ := suite.MarketKeeper.CreateNewDeal(suite.Ctx, owner, tempDeal)
+	suite.Require().Equal(dealId, uint64(2))
+	deal, _ := suite.MarketKeeper.GetDeal(suite.Ctx, 2)
```
Comment (gyuguen) [inline x/market/keeper/deal_test.go:54]:
```suggestion
	deal, _ := suite.MarketKeeper.GetDeal(suite.Ctx, dealId)
```
I think it not good to magicNumber.

File: x/market/keeper/deal_test.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		Budget:               &sdk.Coin{Denom: "umed", Amount: sdk.NewInt(10000000)},
+		WantDataCount:        10000,
+		TrustedDataValidator: []string{acc2.String()},
+		Owner:                acc1.String(),
+	}
+
+	owner, _ := sdk.AccAddressFromBech32(tempDeal.GetOwner())
+
+	dealId, _ := suite.MarketKeeper.CreateNewDeal(suite.Ctx, owner, tempDeal)
+	suite.Require().Equal(dealId, uint64(2))
```
Comment (gyuguen) [inline x/market/keeper/deal_test.go:53]:
```suggestion
	suite.Require().Equal(dealId, expectedDealId)
```
I think we should let them know that the value 2 is the expected value.

File: x/market/types/deal_test.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package types_test
+
+import (
+	"fmt"
+	"github.com/medibloc/panacea-core/v2/x/market/types"
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+func TestNewDealAddress(t *testing.T) {
```
Comment (gyuguen) [inline x/market/types/deal_test.go:10]: Is this test that the same two deal_ids return the same address?

File: x/market/keeper/deal.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+		authtypes.NewBaseAccountWithAddress(
+			sdk.AccAddress(newDeal.GetDealAddress()),
+		),
+		newDeal.DealAddress),
+	)
+	k.accountKeeper.SetAccount(ctx, acc)
+
+	err = k.bankKeeper.SendCoins(ctx, owner, sdk.AccAddress(newDeal.GetDealAddress()), coins)
+	if err != nil {
+		return 0, nil
```
Comment (gyuguen) [inline x/market/keeper/deal.go:40]:
```suggestion
		return 0, err
```
I think the second return will be an err.
And wrapping error message with `The owner's insufficient balance.`

File: x/market/keeper/deal_test.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+	market.InitGenesis(suite.Ctx, suite.MarketKeeper, types.GenesisState{Deals: map[uint64]*types.Deal{
+		testDeal.GetDealId(): &testDeal,
+	}, NextDealNumber: 2})
+
+	suite.MarketKeeper.SetDeal(suite.Ctx, testDeal)
+
+	err := suite.BankKeeper.AddCoins(suite.Ctx, acc1, defaultFunds)
+	if err != nil {
+		panic(err)
+	}
```
Comment (gyuguen) [inline x/market/keeper/deal_test.go:38]: Later, It seems necessary to test that the owner's balance is insufficient. So this code is move to `TestCreateNewDeal` method.

File: x/market/util/conv/doc.go [hunk 9]
```diff
@@ -0,0 +1 @@
+package conv
```
Comment (gyuguen) [inline x/market/util/conv/doc.go:1]: What is this file?

**Review by gyuguen — CHANGES_REQUESTED** [review]
Thank you for your hard work. I left some revies.

File: x/market/keeper/deal.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+	acc = k.accountKeeper.NewAccount(ctx, authtypes.NewModuleAccount(
+		authtypes.NewBaseAccountWithAddress(
+			sdk.AccAddress(newDeal.GetDealAddress()),
+		),
+		newDeal.DealAddress),
+	)
+	k.accountKeeper.SetAccount(ctx, acc)
+
+	err = k.bankKeeper.SendCoins(ctx, owner, sdk.AccAddress(newDeal.GetDealAddress()), coins)
```
Comment (gyuguen) [inline x/market/keeper/deal.go:38]:
```suggestion
	err = k.bankKeeper.SendCoins(ctx, owner, sdk.AccAddressFromBech32(newDeal.GetDealAddress()), coins)
```

I think you used to [here](https://github.com/cosmos/cosmos-sdk/blob/v0.44.5/types/address.go#L158).

This will work normally from v0.43.0 and later versions, so let's think about implementing it in a different way.

File: app/test_helper.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+			abci.RequestInitChain{
+				Validators:      []abci.ValidatorUpdate{},
+				ConsensusParams: simapp.DefaultConsensusParams,
+				AppStateBytes:   stateBytes,
+			})
+	}
+	return app
+}
+
+func FundAccount(bankKeeper bankkeeper.Keeper, ctx sdk.Context, addr sdk.AccAddress, amounts sdk.Coins) error {
```
[the remark this answers is not in the document]
Comment (inchori) [reply app/test_helper.go:37]: I used this method for helping ``TestCreateNewDeal``, but I replaced this method with ``AddCoins`` methods. I removed the method. Thanks for your feedback.

File: x/market/util/conv/doc.go (hunk 9 above)
Comment (inchori) [reply x/market/util/conv/doc.go:1]: The ``doc.go`` file is default file in [Cosmos SDK 0.43](https://github.com/cosmos/cosmos-sdk/blob/3f8596c1955e40ef30e4abcd06f2237d132db3a9/internal/conv/doc.go), but I think it will be not use in our project so I will remove this file. Thank you for your feedback.

File: x/market/client/cli/flags.go [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+	WantDataCount         uint64   `json:"want_data_count"`
+	TrustedDataValidators []string `json:"trusted_data_validators"`
+}
+
+type XCreateDealInputs createDealInputs
+
+type XCreateDealInputsExceptions struct {
+	XCreateDealInputs
+	Other *string
+}
```
Comment (gyuguen) [inline x/market/client/cli/flags.go:25]: I wonder why X is attached to prefix.
Is this auto-generated?

File: x/market/keeper/deal.go [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+	newDeal := &types.Deal{
+		DealId:               dealId,
+		DealAddress:          dealAddress.String(),
+		DataSchema:           deal.GetDataSchema(),
+		Budget:               deal.GetBudget(),
+		TrustedDataValidator: deal.GetTrustedDataValidator(),
+		WantDataCount:        deal.GetWantDataCount(),
+		CompleteDataCount:    0,
+		Owner:                deal.GetOwner(),
+		Status:               "ACTIVE",
```
Comment (gyuguen) [inline x/market/keeper/deal.go:63]: I think it would be better set `const`

```go
const (
  ACTIVE      = "ACTIVE"
  DEACTIVE    = "DEACTIVE"
  COMPLETED    = "COMPLETED"
)
```

File: x/market/client/cli/tx.go [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+		return nil, err
+	}
+
+	err = deal.UnmarshalJSON(contents)
+	if err != nil {
+		return nil, err
+	}
+
+	return deal, nil
+}
```
Comment (hansol-medi) [inline x/market/client/cli/tx.go:71]: How about to move these functions to `x/market/client/cli/txDeal.go` like `did` module?

File: x/market/client/cli/tx.go [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+func parseCreateDealFlags(fs *flag.FlagSet) (*createDealInputs, error) {
+	deal := &createDealInputs{}
+	dealFile, _ := fs.GetString(FlagDealFile)
+
+	if dealFile == "" {
+		return nil, fmt.Errorf("need deal json file using --%s flag", FlagDealFile)
+	}
+
+	contents, err := ioutil.ReadFile(dealFile)
```
Comment (hansol-medi) [inline x/market/client/cli/tx.go:60]: Does it mean the `CreateDealInput` is received as a file?
If then, we'd better receive it as args, not file.

File: proto/panacea/market/v2/deal.proto [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+// Deal defines a deal.
+// TODO: Could remove complete_data_count field
+message Deal {
+  uint64 deal_id = 1;
+  string deal_address = 2;
+  repeated string data_schema = 3;
+  cosmos.base.v1beta1.Coin budget = 4;
+  repeated string trusted_data_validator = 5;
+  uint64 want_data_count = 6;
+  uint64 complete_data_count = 7;
```
Comment (hansol-medi) [inline proto/panacea/market/v2/deal.proto:18]: These two variable names are a bit confusing.
How about `target_num_data` and `filled_num_data`?

**Review by hansol-medi — COMMENTED** [review]
Thank you for your works!
I left some comments.

File: x/market/keeper/deal.go (hunk 13 above)
Comment (inchori) [reply x/market/keeper/deal.go:63]: Yes, I agree about this. I'll status field to ```const```.

File: x/market/types/deal.go [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+	if len(bz) == 0 {
+		return sdkerrors.Wrap(sdkerrors.ErrUnknownAddress, "address cannot be empty")
+	}
+
+	if len(bz) > address.MaxAddrLen {
+		return sdkerrors.Wrapf(sdkerrors.ErrUnknownAddress, "address max length is %d, got %d", address.MaxAddrLen, len(bz))
+	}
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (inchori) [reply x/market/types/deal.go:51]: The Cosmos SDK v0.42 ```AccAddressFromBech32``` is different from Comsos SDK 0.43 ```AccAddressFromBech32```. The max length of address which is constant variable ```AddrLen``` in v0.42 is 20, but the max length of address in v0.43 is 255. When ```dealAddress``` use v0.42 ```AccAddressFromBech32```, it makes error. So I need to use v0.43 ```AccAddressFromBech32``` for make ```dealAddress```. I'll write comment on this function.

File: x/market/client/cli/tx.go (hunk 15 above)
Comment (inchori) [reply x/market/client/cli/tx.go:60]: Yes, I've heard that when we create deal, we will use json file.

File: x/market/client/cli/tx.go (hunk 14 above)
Comment (inchori) [reply x/market/client/cli/tx.go:71]: Yes, I agree with your feed back. Thank you!

File: x/market/client/cli/tx.go (hunk 15 above)
Comment (gyuguen) [reply x/market/client/cli/tx.go:60]: I think it's no proble to receive data in a file. 
Because the [update of did](https://github.com/medibloc/panacea-core/blob/master/x/did/client/cli/txDID.go#L100) and [pool creation of osmosis](https://github.com/osmosis-labs/osmosis/blob/main/x/gamm/client/cli/tx.go#L46) is already received as a json file.

However,some tx uses both json file and argument.

I think it would be good to apply in both ways.(json file and argument) 😀

File: x/market/keeper/deal.go [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+	k.accountKeeper.SetAccount(ctx, acc)
+
+	err = k.bankKeeper.SendCoins(ctx, owner, dealAddress, coins)
+	if err != nil {
+		return 0, sdkerrors.Wrapf(types.ErrNotEnoughBalance, "The owner's balance is not enough to make deal")
+	}
+	return newDeal.GetDealId(), nil
+}
+
+func (k Keeper) newDeal(dealId uint64, deal types.Deal) (types.Deal, error) {
```
[the remark this answers is not in the document]
Comment (gyuguen) [reply x/market/keeper/deal.go:50]: Oh that's good. This function is only used within the keeper.

File: x/market/client/cli/flags.go (hunk 12 above)
Comment (inchori) [reply x/market/client/cli/flags.go:25]: Oh, The prefix X is not auto-generated. I refer to [Osmosis's parse.go](https://github.com/osmosis-labs/osmosis/blob/main/x/gamm/client/cli/parse.go) and I just use X prefix.

File: proto/panacea/market/v2/deal.proto (hunk 16 above)
Comment (youngjoon-lee) [reply proto/panacea/market/v2/deal.proto:18]: Agree. We can have better names. Also, as Gyuguen said, I'm curious what the `complete_data_count` is for. Let me review other codes first.

File: x/market/client/cli/flags.go [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+const (
+	FlagDealFile = "deal-json-file"
+)
+
+type createDealInputs struct {
+	DataSchema            []string `json:"data_schema"`
+	Budget                string   `json:"budget"`
+	WantDataCount         uint64   `json:"want_data_count"`
+	TrustedDataValidators []string `json:"trusted_data_validators"`
+}
```
Comment (gyuguen) [inline x/market/client/cli/flags.go:18]: How about using `MsgCreateDeal` in `proto/panacea/market/v2/tx.proto`.

If Budget is a problem, it seems good to change a spec.

And you can refer to the `json_name` option of  `did/did.proto`

File: x/market/client/cli/txDeal.go [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+
+			return tx.GenerateOrBroadcastTxWithFactory(clientCtx, txf, msg)
+		},
+	}
+
+	cmd.Flags().AddFlagSet(FlagSetCreateDeal())
+	flags.AddTxFlagsToCmd(cmd)
+
+	_ = cmd.MarkFlagRequired(FlagDealFile)
+
```
Comment (gyuguen) [inline x/market/client/cli/txDeal.go:37]:
```suggestion
	cmd.Flags().String(FlagDealFile, "" ,"Deal json file path")
	_ = cmd.MarkFlagRequired(FlagDealFile)

	flags.AddTxFlagsToCmd(cmd)

```

How about fix it like this?
I think we don't need a `market/client/cli/flags.go` files.

File: x/market/client/cli/flags.go [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+	if err := decoder.Decode(&createDealEX); err != nil {
+		return nil
+	}
+
+	*input = createDealInputs(createDealEX.XCreateDealInputs)
+	return nil
+}
+
+func FlagSetCreateDeal() *flag.FlagSet {
+	fs := flag.NewFlagSet("", flag.ContinueOnError)
```
Comment (gyuguen) [inline x/market/client/cli/flags.go:41]: What does this settings mean?
I wonder what changes without this setting.

File: x/market/keeper/deal.go [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+		return 0, err
+	}
+
+	var coins sdk.Coins
+	coins = append(coins, *deal.GetBudget())
+
+	acc := k.accountKeeper.GetAccount(ctx, sdk.AccAddress(newDeal.GetDealAddress()))
+	if acc != nil {
+		return 0, sdkerrors.Wrapf(types.ErrDealAlreadyExist, "deal %d already exist", dealId)
+	}
```
Comment (youngjoon-lee) [inline x/market/keeper/deal.go:26]: It seems that this logic is only for checking if the deal module account already exist. Did I understand it correctly?

Then, I think we need to use `AccDealAddressFromBech32(newDeal.GetDeadAddress())` (or something similar).

File: x/market/types/deal.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (hansol-medi) [reply x/market/types/deal.go:51]: I see. The dealAddress is generated by hash function, which is implemented in v0.43.x, with its dealId.
The resulting address may exceed 20 bytes. 
Is this the way osmosis did for create pool?

**Review by youngjoon-lee — COMMENTED** [review]
Thank you for investigating how to use module accounts. I guessed that the most challenging part is the module account.
I left some comments, but didn't review all codes yet. I will revisit this PR soon.

File: x/market/types/deal.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (inchori) [reply x/market/types/deal.go:51]: Yes, Osmosis also use this function for making pool address.

File: x/market/keeper/deal.go (hunk 22 above)
Comment (youngjoon-lee) [reply x/market/keeper/deal.go:26]: I'm not sure since I'm not familiar with module accounts. Please check if the code above works well as you indended.

File: x/market/types/deal.go [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+		return nil, err
+	}
+
+	return sdk.AccAddress(bz), nil
+}
+
+func VerifyDealAddressFormat(bz []byte) error {
+	verifier := sdk.GetConfig().GetAddressVerifier()
+	if verifier != nil {
+		return nil
```
Comment (youngjoon-lee) [inline x/market/types/deal.go:39]:
```suggestion
		return verifier(bz)
```
Shouldn't we call the `verifier` function? as https://github.com/cosmos/cosmos-sdk/blob/33dbf6a795e68c744a2c312b76d41614b3563c01/types/address.go#L143

File: x/market/types/deal.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (youngjoon-lee) [reply x/market/types/deal.go:51]: I just understood why this logic needs to be copied from cosmos-sdk. As we discussed, this workaround is what Osmosis did: https://github.com/osmosis-labs/osmosis/pull/169. Thanks for digging into this issue.

Please mark these functions or files clearly (e.g. `v044_temp`), so that we can remove them after upgrading cosmos-sdk to v0.44.

File: proto/panacea/market/v2/deal.proto (hunk 16 above)
Comment (youngjoon-lee) [reply proto/panacea/market/v2/deal.proto:18]: It seems that `complete_data_count` is not used anywhere. If there is no special reason to have this field, we can remove it.

If you think this field should be preserved for future use, please let me know.

File: x/market/types/deal.go (hunk 23 above)
Comment (inchori) [reply x/market/types/deal.go:39]: I agree with your review. I'll replace ```nil``` with ```verfiier(bz)```.

File: x/market/client/cli/flags.go (hunk 12 above)
Comment (youngjoon-lee) [reply x/market/client/cli/flags.go:25]: IMHO, we don't need `XCreateDealInputsExceptions`. I just looked at Osmosis source codes, and it seems that the `Other` field is for copying properties of an other existing pool in order to create a new pool. In our case, we aren't gonna need this `Other` option.

File: x/market/client/cli/flags.go (hunk 21 above)
Comment (youngjoon-lee) [reply x/market/client/cli/flags.go:41]: I guess this is the general pattern to create an unnamed flag set using https://github.com/spf13/pflag (if we don't need any name for the flag set).

Comment (youngjoon-lee) [reply x/market/client/cli/flags.go:41]: Anyway, in my opinion, we don't need to follow the Osmosis' way. 

For consistency, I'd suggest to define flags one by one using: `cmd.Flags().String(...)`: https://github.com/medibloc/panacea-core/blob/388585acad2058b5de0c98ef615f8ed41eb286b6/x/aol/client/cli/txTopic.go#L42

When upgrading cosmos-sdk to v0.44, we can change all modules to use `FlagSet`.

File: x/market/client/cli/flags.go (hunk 19 above)
Comment (youngjoon-lee) [reply x/market/client/cli/flags.go:18]: IMHO, it's ok to have this separated `createDealInputs` struct, because this struct is slightly different from `MsgCreateDeal`. I don't think it's good to modify the `tx.proto` for adding some json_name tags.

Also, `MsgCreateDeal` can be more complex in the future, as [MsgCreatePool](https://github.com/osmosis-labs/osmosis/blob/ba4dc51cc8e5c45e96fa14fd4b5ca9321ca92a78/x/gamm/types/tx.pb.go#L34) in the Osmosis.

File: x/market/client/cli/flags.go (hunk 21 above)
Comment (gyuguen) [reply x/market/client/cli/flags.go:41]: Oh, This create a flagset without any name. 
I understand. 😀

**Comment by youngjoon-lee** [comment]
Reviewed all files. Please ping me after resolving (or rejecting) feedbacks above. Then, I'll revisit.

File: x/market/client/cli/flags.go (hunk 19 above)
Comment (gyuguen) [reply x/market/client/cli/flags.go:18]: Aha, It's ok.
Let's leave this struct as it is.

File: x/market/client/cli/flags.go (hunk 21 above)
Comment (inchori) [reply x/market/client/cli/flags.go:41]: I agree with that we need to follow Osmosis' way and not yet to use ```FlagSet```. Thanks for your feedback.

File: x/market/client/cli/txDeal.go (hunk 20 above)
Comment (inchori) [reply x/market/client/cli/txDeal.go]: I think this review is related to below issue. So I will resolve this conversation. Thank you for your feedback 😀

File: x/market/keeper/deal.go (hunk 22 above)
Comment (inchori) [reply x/market/keeper/deal.go:29]: Yes, This is my mistake. I didn't change this code yet. I'll change it to ```AccDealAddressFromBech32(newDeal.GetDealAddress())```. Thank you for your feedback.

File: x/market/client/cli/flags.go [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+	dec := json.NewDecoder(bytes.NewReader(data))
+	dec.DisallowUnknownFields()
+
+	if err := dec.Decode(&createDeal); err != nil {
+		return nil
+	}
+
+	*input = createDealInputs(createDeal)
+	return nil
+}
```
Comment (gyuguen) [inline x/market/client/cli/flags.go:32]: How about moving all of thes sources to `x/market/client/cli/txDeal.go`?

These don't seem to be related to `flag`

Comment (inchori) [reply x/market/client/cli/flags.go:32]: Thank your for your review. I think it's good idea but about rename this file to ```parse.go```?

File: x/market/keeper/deal_test.go [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+	return types.Deal{
+		DealId:                1,
+		DealAddress:           types.NewDealAddress(1).String(),
+		DataSchema:            []string{acc1.String()},
+		Budget:                &sdk.Coin{Denom: "umed", Amount: sdk.NewInt(1000000000)},
+		TargetNumData:         10000,
+		FilledNumData:         0,
+		TrustedDataValidators: []string{acc2.String()},
+		Owner:                 acc1.String(),
+		Status:                "ACTIVE",
```
Comment (gyuguen) [inline x/market/keeper/deal_test.go:111]: Please change to const.

```suggestion
		Status:                ACTIVE,
```

File: proto/panacea/market/v2/deal.proto [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+// Deal defines a deal.
+// TODO: Could remove complete_data_count field
+message Deal {
+  uint64 deal_id = 1;
+  string deal_address = 2;
+  repeated string data_schema = 3;
+  cosmos.base.v1beta1.Coin budget = 4;
+  repeated string trusted_data_validators = 5;
+  uint64 target_num_data = 6;
+  uint64 filled_num_data = 7;
```
Comment (youngjoon-lee) [inline proto/panacea/market/v2/deal.proto:18]: Aha. Now I understand what this field means. I guess you're gonna update this value whenever you get a SellData tx. 
Then, I think the `cur num data` and `max num data` would be clearer. wdyt?
it's little awkward that the deal is 'filled'. 

btw. i'm not sure yet how we'll improve this deal concept in near future. But, from my gut feeling, it would be enough to have only budget and pricePerData. Then, the deal becomes deactivated when the balance of the deal account runs out.
But, don't think too much of this now. let's see when we implement SellData tx

Comment (inchori) [reply proto/panacea/market/v2/deal.proto:18]: I agree with your review that rename the field to ```cur_num_data``` and ```max_num_data```.  Also we need to think this when we implement SellData feature. And It's good idea that we just have ```budget``` field and ```pricePerData``` field! Thank you for your review.

Comment (hansol-medi) [reply proto/panacea/market/v2/deal.proto:18]: Good! `max_num_data` and `cur_num_data` are clearer.

**Review by gyuguen — APPROVED** [review]
LGTM 👍 
I checked all the reviews
Please check youngjoon-lee commants.

File: x/market/types/deal.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (inchori) [reply x/market/types/deal.go]: This review is quite important topic for create-deal feature. So I will not resolve this conversation.

**Review by youngjoon-lee — APPROVED** [review]
Thank you! WAGMI!!

**Review by hansol-medi — APPROVED** [review]
👍👍👍
